rfcs/002 - Compiler Refactor RFC.md

```diff
@@ -0,0 +1,141 @@
+5. prompt LLM to generate all functions and classes based on output from (3) or error (only if LLM unreachable), generating 3 different files by default to improve reliability for a greater expense
+6. prompt LLM to generate test suite for all functions based on output from (3) or error (can be done in parallel when using ChatGPT), generating 3 different files by default to improve reliability for a greater expense
+7. take pairs of output from (5) and (6) and verify they are valid markdown with the expected sections only, returning the pair(s) that work, or erroring if none do
+8. take each pair of outputs from (5) and (6) and run them through a Python linter to confirm they are valid python (most stylistic rules removed, just syntax, unused imports, and a few others), if it errors it sends all of this to an LLM to generate new versions of the file, repeating on error until a limit is reached, then exiting the flow for that particular pair, erroring if all pairs fail.
+9. take each pair of outputs from (8) and runs the test suite, passing them and the test suite errors to an LLM if they fail, otherwise continuing. If it continues to fail for too long, the particular pair exits the flow, if all pairs exit the flow, it errors.
+10. the first pair to finish (9) cancels the rest of the operations, then by default it appends the CLI and HTTP server helper logic to the code file and exits.
+
+For now, *most* of the Marsha language functionality exists within (3), and it is a synchronous operation, but most of the solutions for database support would at minimum require (3) to become async and involve various DB and ML operations (likely: connect to the DB and generate a schema dump, then chunk that dump and index it with vectors, then use vector search to find the most relevant table(s), stored procedure(s), etc to provide to (5) and (6)) but it's possible these features could "bleed" into the validation steps of Marsha, and the codebase starts turning (even more) into a big ball of mud. (For instance, validating the generated SQL DB access code in (5) against the actual DB schema in (7) seems likely).
+
+There are other dimensions we would like to extend Marsha on:
+
+* Which LLM is used to do the work.
+* Which target language Marsha compiles to.
+* Dev tools (debugging, decompiling, etc) for Marsha.
+
+The solution we come up with needs to keep the complexity of all of these different features from compounding on one another, ideally.
+
+#### Proposal: Marsha as Extensions that append to lists of layers
+
+Taking a very rough inspiration from LLMs themselves, we'll have a singular kind of "transformer" that is the foundational piece of the compiler, and they will be organized into layers where each layer executes in parallel and their outputs are fed into the next layer, etc. Each logical feature will be an extension that concatenates its logic into the appropriate layers, with the "meaning" said layers being a convention per extension "group".
```

Comment:
Maybe "transformation" or "mapper" instead of "transformer"

Reply:
Mapper! Of course!